apis/v1alpha2/nginxproxy_types.go

```diff
@@ -412,6 +418,54 @@ type DaemonSetSpec struct {
 	Container ContainerSpec `json:"container"`
 }
 
+// +kubebuilder:validation:XValidation:message="at least one metric must be specified when autoscaling is enabled",rule="!self.enabled || (has(self.targetCPUUtilizationPercentage) || has(self.targetMemoryUtilizationPercentage) || (has(self.autoscalingTemplate) && size(self.autoscalingTemplate) > 0))"
+// +kubebuilder:validation:XValidation:message="minReplicas must be less than or equal to maxReplicas",rule="self.minReplicas <= self.maxReplicas"
+// +kubebuilder:validation:XValidation:message="CPU utilization must be between 1 and 100",rule="!has(self.targetCPUUtilizationPercentage) || (self.targetCPUUtilizationPercentage >= 1 && self.targetCPUUtilizationPercentage <= 100)"
+// +kubebuilder:validation:XValidation:message="memory utilization must be between 1 and 100",rule="!has(self.targetMemoryUtilizationPercentage) || (self.targetMemoryUtilizationPercentage >= 1 && self.targetMemoryUtilizationPercentage <= 100)"
+//
+//nolint:lll
+type HPASpec struct {
+	// behavior configures the scaling behavior of the target
+	// in both Up and Down directions (scaleUp and scaleDown fields respectively).
+	// If not set, the default HPAScalingRules for scale up and scale down are used.
+	//
+	// +optional
+	Behavior *autoscalingv2.HorizontalPodAutoscalerBehavior `json:"behavior,omitempty"`
+
+	// AutoscalingTemplate configures the additional scaling option.
+	//
+	// +optional
+	AutoscalingTemplate *[]autoscalingv2.MetricSpec `json:"autoscalingTemplate,omitempty"`
```

Comment:
I wonder if we just name this field `Metrics`, and then remove the TargetCPUand TargetMemory fields? Seems like we are defining specific metrics types when we could just keep it generic and a user defines whatever they want.

Reply:
> I wonder if we just name this field `Metrics`, and then remove the TargetCPUand TargetMemory fields? Seems like we are defining specific metrics types when we could just keep it generic and a user defines whatever they want.

@sjberman Yes, we can do what you mentioned.
However, it's common to configure CPU and memory for HPA separately. while using the AutoscalingTemplate for custom or additional metrics. This approach is clean and flexible. What do you think?

Reply:
Ok, if you feel that these are common enough things to configure, then making them first-class configuration options is fine with me.